src/group_theory/conjugates.lean

```diff
@@ -0,0 +1,142 @@
+/-
+
+Author: Michael Howes and Scott Morrison, 2019
+
+The definition of conjugates in a group and the definition
+of the normal closure of a subset of a group.
+-/
+import algebra.group
+import group_theory.subgroup
+import tactic.rewrite
+
+variable {α : Type}
+variable [group α]
+
+
+
+/-- Two elements a b of a group are conjugates if there exists a
+c such that c*a*c⁻¹ = b -/
+def is_conjugate (a b : α) : Prop := ∃ c : α, c*a*c⁻¹ = b
+
+/-- Some basic properties of conjugates -/
+@[simp] lemma conj_inv {a b : α} : (b*a*b⁻¹)⁻¹ = b*a⁻¹*b⁻¹ :=
+begin
+  rw [mul_inv_rev _ b⁻¹, mul_inv_rev b _, inv_inv, mul_assoc],
+end
+
+@[simp] lemma conj_mul {a b c : α} : (b*a*b⁻¹)*(b*c*b⁻¹) = b*(a*c)*b⁻¹ :=
+begin
+  assoc_rw inv_mul_cancel_right,
+  repeat {rw mul_assoc},
+end
+
+namespace is_conjugate
+/-- is_conjugate is an equivalence relation -/
+lemma refl {a : α} : is_conjugate a a := exists.intro (1:α) (by simp)
+
+lemma symm_aux {a b c : α} : c*a*c⁻¹ = b → a=c⁻¹*b*c  :=
+λ h, by rw [mul_assoc, ←(eq_inv_mul_of_mul_eq (eq_mul_of_mul_inv_eq h))]
+
+lemma symm {a b : α} : is_conjugate a b → is_conjugate b a :=
+λ ⟨c,h⟩, ⟨c⁻¹, by simpa only [inv_inv] using eq.symm (symm_aux h)⟩
+
+lemma trans_aux {a c d e : α} : d * a * d⁻¹ = e⁻¹ * c * e → e*d*a*(d⁻¹*e⁻¹) = c :=
+λ h,
+begin
+  assoc_rw h,
+  simp,
+end
+
+lemma trans {a b c : α} : is_conjugate a b → is_conjugate b c → is_conjugate a c :=
+λ ⟨d,w⟩ ⟨e,h⟩,
+begin
+  rw (symm_aux h) at w,
+  exact ⟨e*d, by simpa only [mul_inv_rev] using (trans_aux w)⟩
+end
+
+end is_conjugate
+
+/-- Here the normal closure of a set is defined as the subgroup closure of the
+set of conjugates of the given conjugates -/
+
+variables {s : set α}
+
+/-- Given an element a, conjugates a is the set of conjugates -/
+def conjugates (a : α) : set α := is_conjugate a
+
+lemma self_mem_conjugates {a : α} : a ∈ conjugates a := is_conjugate.refl
+
+/-- Given a set s, conjugates_of_set s is the set of all conjugates of
+the elements of s. -/
+def conjugates_of_set (s : set α) : set α :=
```

Comment:
This can be `⋃ a ∈ s, conjugates a`

Reply:
I rewrote this `{b | {b | ∃ a ∈ s, is_conj a b}`. Writing the set as union required a lot of changes to my other proofs.

Reply:
How bad are the changes? The definition Mario proposed is certainly more readable.

Reply:
Mario's definition also has the benefit that someone who wants to use your definition has access to all the library functions about `⋃ a ∈ s, blah`.

Reply:
The main reason why the changes are difficult is because I use anonymous constructors in many of my proves. With the definition I currently have of `conjugates_of_set`, a hypothesis of the form `h : x ∈ conjugates_of_set s` can be written as an anonymous constructor  `⟨a, H, h⟩` where `a : α`, `H : a ∈ s` and `h : is_conj a x`. I think it might take a while to rewrite the proofs using the union definition.

That being said it is a much easier to parse definition so I can make the change if there is demand.

Reply:
You can still use anonymous constructors with the union definition, it's just a little messier, c.f. `⟨a, ⟨_, H, rfl⟩, h, rfl⟩` instead (or something similar). Actually, this brings up a point about the not-nice definition of indexed union, but it's independent of this PR.

Reply:
Okay, I will try to rewrite what I've done.

Reply:
It is now a union.